### PR TITLE
Add initial abstract audio and sound classes

### PR DESCRIPTION
**Pull request by docEdub**
Adds the initial set of abstract audio and sound classes to be used for Audio Engine v2.

### CODE REVIEW

File: packages/dev/core/src/Audio/v2/abstractAudioNode.ts [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+     * Undefined for input nodes.
+     */
+    protected _connectedDownstreamNodes?: Array<AbstractAudioNode>;
+
+    /**
+     * The connected upstream audio nodes.
+     *
+     * Undefined for output nodes.
+     */
+    protected _connectedUpstreamNodes?: Array<AbstractAudioNode>;
```
Comment (docEdub) [inline packages/dev/core/src/Audio/v2/abstractAudioNode.ts:128]: Is it ok for one of these to be `undefined` for the lifetime of the `AbstractAudioNode` object, or is that strange? I was thinking maybe they should be `Nullable` instead, but it would be `null` for the node's entire lifetime if we did that, and we probably shouldn't be adding a property that will never be used, right?

Comment (docEdub) [reply packages/dev/core/src/Audio/v2/abstractAudioNode.ts:128]: We could break this class up into separate classes based on the 3 node types (`Input`, `Output`, and `InputOutput`), but the class hierarchy gets difficult to work with if we do that.

File: packages/dev/core/src/Audio/v2/IAudioNodeParent.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+/* eslint-disable babylonjs/available */
+/* eslint-disable jsdoc/require-jsdoc */
+
+import type { AbstractAudioNode } from "./abstractAudioNode";
+import type { IDisposable } from "../../scene";
+
+export interface IAudioNodeParent extends IDisposable {
+    _addChildNode(node: AbstractAudioNode): void;
```
[the remark this answers is not in the document]
Comment (ryantrem) [reply packages/dev/core/src/Audio/v2/IAudioNodeParent.ts:8]: Getting further into this PR, I think I understand the intent. This is only "public" so that classes that implement this interface (specifically `AbstractAudioNode`) can call `_addChildNode` when their `parent` is set, correct? It's not actually intended to be public in the API, yes? If that is the case, then you could use an abstract base class that exposes protected static functions for adding/removing child nodes. This way it keeps these out of the public contract, but still allows derived classes to mutate the child nodes. I created a quick TS Playground [example here](https://www.typescriptlang.org/play/?target=7#code/PTCWFsAcHsCcBcAEBvACrakDOB1U8ALAYQNABsATWAUwDsAaRI6KaWu+AFQE9JrGAYgFdaAY3ig2zVu1rxGQrNSIBDMmQBGK0QGtG4auGgBfRADMM4RACIa2+NYDcAKBAQYCFKjIr4ZuOCMAMrw3GTU6JiMAGqg1ADuIWH8iAAyKtzQQvAkKrQA5tQAogBuHDFx8Yzo1FhYKhrhFQmR2NU0dQ3hIb7KapraOjx8phYsNnbiALS0vqBlTs7OomxY0OEAdKLhKrAAFACULs4NWPCw9ojbKnWIAIIaZxfid0IUkgBy0BQRuxwozkQQMQkFg816iAA+qJSJQvj8sAB+ABciCC1HgAB4Hk97K93tB4dQAHzHYEgjDwajiagUKEqCgUEjkChEva0b7UVE4854t6fTkHAHk8mgMyIPYAQkIoCwG2hsNZnKwQuQgJFGplcoVLKJWEQAF5EOx4miMYcXBqgcYllbEFr5TDdcqNgyKOzBZbgTb1UDQdAqTS6ZCaEYysy4ZyPT9uY9eS9+YTBaiStBQHS1XaHTrIwjERsfuEqdHqEdfYgfeT-YGqXSznNRPTGRGlT89pA-nJY7iEwSiahO-JjZzu-H4PiBT9VeWq4P5W6W2yOVOvdbbcDq9Ta4h6xJGyHDNBw4q2R2aF37nHnuPE-3B4xl1zLz2b33BcKrWeOPLQ0flCeo0fMtyRtH1TjHK4fFuHlrwnJMfkQagAA8qVoCh9Rgvk3x+AdzyQTMNyERpQEbOwKDYMhuEQgpQHYUdYMTIoaPYQ1jSEdRJVXCkAy3WkoS-OQUWfMc4LvPCyUI4jG0KJABPgQ56Kwydfjwj8NRoeAhFgWh7VIbU5MQRFEV02UNjofJaOoLjK0ksgSJ3DEQUHdtB0U3tlNw-4AB9EBEH4zEsihpztMUJWzAyDUipy8OCu1gQ0rTaC4kD1w1TD3PglTvwPMN-2dNsHTkxgtWAq1wsHVi5OS4F0tfDy50hBcAIKvSNiKkyVWs5wwKvS5rmg3qMqYiyWOQ1D0OEhjsKyuQAW6pYi2HGNJqUzLWNodiyE41xgEQL4KXBKkoTMERxEkWh9WQmAlAoZxHw2ZwgA).

File: packages/dev/core/src/Audio/v2/abstractAudioNode.ts [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+
+    public engine: AbstractAudioEngine;
+
+    // If parent is null, node is owned by audio engine.
+    protected _parent: Nullable<IAudioNodeParent> = null;
+
+    public get parent(): IAudioNodeParent {
+        return this._parent ?? this.engine;
+    }
+
```
Comment (ryantrem) [inline packages/dev/core/src/Audio/v2/abstractAudioNode.ts:80]: Seems like this should just be the setter side of the property rather than a function.
```suggestion
  public set parent(parent: Nullable<IAudioNodeParent>) {
```

File: packages/dev/core/src/Audio/v2/abstractAudioNode.ts (hunk 1 above)
Comment (sebavan) [reply packages/dev/core/src/Audio/v2/abstractAudioNode.ts:128]: It is totally fine for me but the type needs to be changed to `Array<AbstractAudioNode> | undefined`

Comment (ryantrem) [reply packages/dev/core/src/Audio/v2/abstractAudioNode.ts:128]: Curious @sebavan, why explicitly add the `| undefined`? It's already implicit from the `?`. I'm mostly curious because I would have wrote this code the same way @docEdub did in this case. I would expect either `?` and no explicit `| undefined`, or no `?` and an explicit `| undefined` (although the two are not semantically the same).

Comment (docEdub) [reply packages/dev/core/src/Audio/v2/abstractAudioNode.ts:128]: I was not aware of this before, but if I'm reading [the `exactOptionalPropertyTypes` docs](https://www.typescriptlang.org/tsconfig/#exactOptionalPropertyTypes)  right, then I **don't** want `Array<AbstractAudioNode> I undefined` because this would allow setting it to `undefined` after the constructor is called. Is that right? So, if I don't want these properties to be set after the constructor is called, then I should leave it as-is, with just the `?`.

Thinking more on this, these properties can be readonly, or private with a protected getter.

File: packages/dev/core/src/Audio/v2/abstractAudioNode.ts (hunk 3 above)
Comment (docEdub) [reply packages/dev/core/src/Audio/v2/abstractAudioNode.ts:80]: Ya, I went back and forth on this one. I think it makes more sense to use a setter, too, but I'm thinking we can add optional arguments later with `setParent(...)` and still keep backward compatibility. I'm not sure which way to go, yet.

File: packages/dev/core/src/Audio/v2/abstractAudioListener.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+    public get up(): Vector3 {
+        return this._up;
+    }
+
+    public set up(value: Vector3) {
+        this._up.copyFrom(value);
+    }
+
+    private _attachedMesh: Nullable<AbstractMesh> = null;
```
[the remark this answers is not in the document]
Comment (docEdub) [reply packages/dev/core/src/Audio/v2/abstractAudioListener.ts:44]: I'm not sure what to do with this one. `TransformNode` makes the most sense to me, but will users know they can use it for meshes if we do that? Maybe add and `attachTransformNode` function and keep the `attachMesh` function, too?

File: packages/dev/core/src/Audio/v2/abstractAudioListener.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+export abstract class AbstractAudioListener extends AbstractAudioNode {
+    public constructor(parent: AbstractAudioNode) {
+        super(parent.engine, AudioNodeType.InputOutput);
+
+        this.setParent(parent);
+    }
+
+    private _position = Vector3.Zero();
+    private _forward = Vector3.Forward();
+    private _up = Vector3.Up();
```
[the remark this answers is not in the document]
Comment (docEdub) [reply packages/dev/core/src/Audio/v2/abstractAudioListener.ts:18]: Ya, this way matches the WebAudio API, but a Quaternion might be better at this level of the API.

File: packages/dev/core/src/Audio/v2/abstractAudioEngine.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    public _removeChildNode(node: AbstractAudioNode): void {
+        const index = this._childNodes.indexOf(node);
+        if (index < 0) {
+            return;
+        }
+
+        this._childNodes.splice(index, 1);
+    }
+
+    private _soundInstances = new Array<AbstractStaticSoundInstance>();
```
[the remark this answers is not in the document]
Comment (docEdub) [reply packages/dev/core/src/Audio/v2/abstractAudioEngine.ts:43]: I'm not sure I follow. How would a Map be simpler? Would the sound instance be the Map's key, or its value?

File: packages/dev/core/src/Audio/v2/abstractAudioNode.ts [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+    }
+}
+
+export abstract class AbstractNamedAudioNode extends AbstractAudioNode {
+    public constructor(name: string, engine: AbstractAudioEngine, nodeType: AudioNodeType) {
+        super(engine, nodeType);
+        this.name = name;
+    }
+
+    public name: string;
```
[the remark this answers is not in the document]
Comment (docEdub) [reply packages/dev/core/src/Audio/v2/abstractAudioNode.ts:229]: This will probably change later.

File: packages/dev/core/src/Audio/v2/abstractAudioBusNode.ts [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+}
+
+export abstract class AbstractAudioBusNode extends AbstractNamedAudioNode {
+    public constructor(name: string, engine: AbstractAudioEngine, options?: IAudioBusNodeOptions) {
+        super(name, engine, AudioNodeType.InputOutput);
+
+        this.volume = options?.volume ?? 1;
+    }
+
+    public volume: number;
```
[the remark this answers is not in the document]
Comment (docEdub) [reply packages/dev/core/src/Audio/v2/abstractAudioBusNode.ts:18]: This will change later.

File: packages/dev/core/src/Audio/v2/abstractAudioDevice.ts [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+export abstract class AbstractAudioDevice extends AbstractNamedAudioNode {
+    public constructor(name: string, engine: AbstractAudioEngine, options?: IAudioDeviceOptions) {
+        super(name, engine, AudioNodeType.Input);
+
+        if (options?.enableListener) {
+            this._listener = engine.createListener(this);
+        }
+    }
+
+    private _listener: Nullable<AbstractAudioListener> = null;
```
[the remark this answers is not in the document]
Comment (docEdub) [reply packages/dev/core/src/Audio/v2/abstractAudioDevice.ts:22]: Possibly. Probably not. I'm still debating if it makes sense to let users enable it after construction.

File: packages/dev/core/src/Audio/v2/abstractAudioEngine.ts [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+    public dispose(): void {
+        for (const node of this._childNodes) {
+            node.dispose();
+        }
+        this._childNodes.length = 0;
+    }
+
+    private _childNodes = new Array<AbstractAudioNode>();
+
+    public _addChildNode(node: AbstractAudioNode): void {
```
[the remark this answers is not in the document]
Comment (docEdub) [reply packages/dev/core/src/Audio/v2/abstractAudioEngine.ts:26]: What would you do with the currently public `AbstractAudioEngine._addSoundInstance`, which will be called from the `AbstractSoundInstance` constructor?

I'm not sure this is a problem since we can just mark it as `@internal`. I see this being done a fair amount already in other classes. AFAIK the compiled JavaScript doesn't have private members anyway, and users are advised to avoid using members with an `_` prefix.

File: packages/dev/core/src/Audio/v2/abstractAudioNode.ts (hunk 3 above)
Comment (docEdub) [reply packages/dev/core/src/Audio/v2/abstractAudioNode.ts:80]: I made it protected instead of public since the user shouldn't need to call it.

File: packages/dev/core/src/Audio/v2/IAudioNodeParent.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (docEdub) [reply packages/dev/core/src/Audio/v2/IAudioNodeParent.ts]: I like the suggestion of adding `AbstractAudioNodeParent`, but I don't like the static functions, yet. I'm not sure the extra code is worth making the TypeScript API cleaner if the resulting JavaScript API stays the same.

File: packages/dev/core/src/Audio/v2/abstractAudioEngine.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (ryantrem) [reply packages/dev/core/src/Audio/v2/abstractAudioEngine.ts]: Sorry I meant Set.

File: packages/dev/core/src/Audio/v2/abstractAudioListener.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (ryantrem) [reply packages/dev/core/src/Audio/v2/abstractAudioListener.ts:42]: I would expect users would understand that a Mesh is a TransformNode. There are other APIs that use the TransformNode level of abstraction.

File: packages/dev/core/src/Audio/v2/abstractAudioEngine.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (Popov72) [reply packages/dev/core/src/Audio/v2/abstractAudioEngine.ts]: Yes, properties/methods that we don't want the user to see but that we want to be able to use internally should be prefixed with `_`, but they should also be marked with `@internal` so that they don't appear in editors when using Intellisense.

**Review by RaananW — COMMENTED** [review]
A few general notes I have:

1) Null assignments - there is no need to assign a variable to null if you don't let the user set its value. It can stay undefined/optional. Use Nullable if the public API allows the user to set it to null
2) Docs - you added the global jsdoc eslint ignore rule on each file - leave it without for now and keep the warnings.
3) getter/setters - if all you do is get a private variable or set the variable to whatever provided, you can avoid getters/setters altogether and use a public member. 
4) order - we don't enforce it, but we do our best to keep class members in a certain order - variables on top (private/protected/public) and functions. It made me wonder if there is an eslint rule for that, which I will search, but for now, try keeping the variables on top of the class before the constructor.

File: packages/dev/core/src/Audio/v2/abstractAudioEngine.ts [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    // Not owned, but all items should in parent's `children` container, too, which is owned.
+    public readonly soundInstances = new Set<AbstractStaticSoundInstance>();
+
+    // Owned
+    public readonly soundSources = new Set<AbstractSoundSource>();
+
+    // Owned
+    private _listeners: Nullable<Set<SpatialAudioListener>> = null;
```
[the remark this answers is not in the document]
Comment (docEdub) [reply packages/dev/core/src/Audio/v2/abstractAudioEngine.ts:48]: Ya, this isn't cleaned up. I'm adding this in `dispose()` ...
```
        this._listeners?.clear();
        this._listeners = null;
```

Does it make sense to set it to `null` in `_removeListener` if the set is empty, to save memory?

File: packages/dev/core/src/Audio/v2/abstractAudioListener.ts [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+import type { Nullable } from "../../types";
+
+export abstract class AbstractAudioListener extends AbstractAudioNode {
+    public constructor(parent: AbstractAudioNode) {
+        super(parent.engine, AudioNodeType.InputOutput, parent);
+    }
+
+    private _position = Vector3.Zero();
+    private _forward = Vector3.Forward();
+    private _up = Vector3.Up();
```
[the remark this answers is not in the document]
Comment (docEdub) [reply packages/dev/core/src/Audio/v2/abstractAudioListener.ts]: This is an unused class I forgot to remove. I decided listeners don't need to be audio nodes since they only need to provide parameters to audio positioner nodes. The `SpatialAudioListener` class replaces this `AbstractAudioListener` node.

File: packages/dev/core/src/Audio/v2/abstractAudioEngine.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (RaananW) [reply packages/dev/core/src/Audio/v2/abstractAudioEngine.ts:48]: yes, the garbage collector will be very thankful :-)

File: packages/dev/core/src/Audio/v2/abstractAudioListener.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (RaananW) [reply packages/dev/core/src/Audio/v2/abstractAudioListener.ts]: Was part of the general note i had regarding getters/setters, which are slightly slower (due to the function call). Avoid them whenever possible and use the if there are some side-effects after setting something.

File: packages/dev/core/src/Audio/v2/abstractAudioBus.ts [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+        }
+    }
+
+    private _positioner: Nullable<AbstractAudioPositioner> = null;
+
+    public get positioner(): Nullable<AbstractAudioPositioner> {
+        return this._positioner;
+    }
+
+    public enablePositioner() {
```
[the remark this answers is not in the document]
Comment (docEdub) [reply packages/dev/core/src/Audio/v2/abstractAudioBus.ts:38]: I'll be adding more things like this later.

File: packages/dev/core/src/Audio/v2/abstractAudioBus.ts [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+        if (this._positioner) {
+            return;
+        }
+
+        this._positioner = this.engine.createPositioner(this);
+    }
+
+    public readonly sender: AbstractAudioSender;
+
+    private _outputBus: Nullable<AbstractMainAudioBus | AbstractAudioBus> = null;
```
[the remark this answers is not in the document]
Comment (docEdub) [reply packages/dev/core/src/Audio/v2/abstractAudioBus.ts]: Thanks! I was wondering about this. Done now

File: packages/dev/core/src/Audio/v2/abstractAudioListener.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (docEdub) [reply packages/dev/core/src/Audio/v2/abstractAudioListener.ts]: Is it considered a back-compat break if we want to change a property to use getters/setters later? I'm thinking it's not a back-compat break since it looks the same in the API. Is that right?

File: packages/dev/core/src/Audio/v2/abstractAudioNode.ts [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+    private _parent: Nullable<AbstractAudioNodeParent> = null;
+
+    public readonly engine: AbstractAudioEngine;
+
+    /**
+     * The connected downstream audio nodes.
+     *
+     * Undefined for input nodes.
+     */
+    protected readonly _connectedDownstreamNodes?: Set<AbstractAudioNode> | undefined;
```
Comment (bghgary) [inline packages/dev/core/src/Audio/v2/abstractAudioNode.ts:36]: This is redundant because of the `?`.

```suggestion
    protected readonly _connectedDownstreamNodes?: Set<AbstractAudioNode>;
```

File: packages/dev/core/src/Audio/v2/abstractAudioNodeParent.ts [hunk 16]
```diff
@@ -0,0 +1,26 @@
+/* eslint-disable babylonjs/available */
+/* eslint-disable jsdoc/require-jsdoc */
+
+import type { AbstractAudioNode } from "./abstractAudioNode";
+import type { Nullable } from "../../types";
+import type { IDisposable } from "../../scene";
+
+export class AbstractAudioNodeParent implements IDisposable {
+    private _children: Nullable<Set<AbstractAudioNode>>;
```
Comment (bghgary) [inline packages/dev/core/src/Audio/v2/abstractAudioNodeParent.ts:9]: I would suggest making this readonly and initializing with a new set so that the rest of code is simpler (i.e., you don't need to check if it exists).

```suggestion
    private readonly _children = new Set<AbstractAudioNode>();
```

File: packages/dev/core/src/Audio/v2/spatialAudioTransform.ts [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    public get position(): Vector3 {
+        this._updatePosition();
+        return this._position;
+    }
+
+    public set position(position: Vector3) {
+        this._position.copyFrom(position);
+    }
```
Comment (bghgary) [inline packages/dev/core/src/Audio/v2/spatialAudioTransform.ts:40]: This pattern is interesting. Do you expect users to be able to call `transform.position.copyFrom()`? Do you have to know when the position is dirty? Maybe the returned Vector3 should be DeepImmutable?

File: packages/dev/core/src/Audio/v2/spatialAudioTransform.ts [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+        this._rotationDirty = true;
+    }
+
+    private _updatePosition() {
+        if (!this._positionDirty) {
+            return;
+        }
+        this._positionDirty = false;
+
+        this._attachedTransformNode!.getWorldMatrix().decompose(undefined, undefined, this._position);
```
Comment (bghgary) [inline packages/dev/core/src/Audio/v2/spatialAudioTransform.ts:87]: This seems problematic. `this._attachedTransformNode` can be null.

File: packages/dev/core/src/Audio/v2/abstractAudioEngine.ts [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+        super.dispose();
+    }
+
+    public get listeners(): Set<SpatialAudioListener> {
+        if (!this._listeners) {
+            this._listeners = new Set();
+        }
+
+        return this._listeners;
+    }
```
Comment (bghgary) [inline packages/dev/core/src/Audio/v2/abstractAudioEngine.ts:56]: Are users of an audio engine expected to be able to modify the set? It seems creating a listener adds to this list, but it's also possible for someone to do this manually? Should the return type be `DeepImmutable`?

Same question for `soundInstances` and `soundSources`.

File: packages/dev/core/src/Audio/v2/abstractAudioSend.ts [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+        return this._sendType;
+    }
+
+    public setSendType(sendType: AudioSendType) {
+        if (this._sendType === sendType) {
+            return;
+        }
+
+        this._sendType = sendType;
+    }
```
Comment (bghgary) [inline packages/dev/core/src/Audio/v2/abstractAudioSend.ts:64]: Why is this a function and not a setter?

**Review by bghgary — COMMENTED** [review]
Sorry for the delayed review. Overall, the structure seems good. I have some concerns with the usage pattern of some of the properties.

File: packages/dev/core/src/Audio/v2/abstractAudioNode.ts (hunk 15 above)
Comment (docEdub) [reply packages/dev/core/src/Audio/v2/abstractAudioNode.ts:36]: Ya, it seems strange to me, too, but we discussed it earlier in https://github.com/BabylonJS/Babylon.js/pull/15483#discussion_r1733224883.

TL;DR: Seb says ...
> https://forum.babylonjs.com/t/support-typescript-tsconfig-flag-exactoptionalpropertytypes-true/33824 will explain the details. We have since received a few PRs to fix it within various places in Babylon.

File: packages/dev/core/src/Audio/v2/spatialAudioTransform.ts (hunk 18 above)
Comment (docEdub) [reply packages/dev/core/src/Audio/v2/spatialAudioTransform.ts:87]: With the way it's implemented right now, `this._attachedTransformNode` will always be non-null if `this._positionDirty` is `true`. I don't think there's a way to let TypeScript know this, though, so I used a `!`.

File: packages/dev/core/src/Audio/v2/spatialAudioTransform.ts (hunk 17 above)
Comment (docEdub) [reply packages/dev/core/src/Audio/v2/spatialAudioTransform.ts:40]: I'm not sure I understand the question here. How and where would users call `transform.position.copyFrom()`?

File: packages/dev/core/src/Audio/v2/abstractAudioNode.ts (hunk 15 above)
Comment (Popov72) [reply packages/dev/core/src/Audio/v2/abstractAudioNode.ts:36]: From the PR that added the flag to Babylon:

> Indeed! They carry slightly different intent with `exactOptionalPropertyTypes: true`: a question mark allows it to be omitted but not set to undefined, while `| undefined` does the opposite. This can help catch a certain class of bugs (e.g., when checking stuff with `”foo” in object` incorrectly).
